### PR TITLE
Ees 2999 speed up choose filters step

### PR DESCRIPTION
**Pull request by duncan-at-hiveit**
This PR was originally to attempt to speed up the Choose Filters step.  Its main contribution however is to speed up the Table Results step considerably by replacing the "SelectObservations" stored procedures used for finding matching Observations with a new query that removes the need for a JOIN with the ObservationFilterItem table in favour of multiple EXISTS clauses.

These new clauses each check that there's at least one matching Filter Item chosen for each available Filter for every Observation row that matches.  Because they're applied one after another, the first EXISTS clause can narrow down the number of rows that the next EXISTS clause needs to check, and so on for each subsequent EXISTS clause.

We make use of a new library in order to generate temporary tables and bulk insert into them too.  

This PR:
* Merges SubjectMetaQueryContext and ObservationQueryContext together - only ObservationQueryContext remains now.
* Replaces the SelectObservation stored procedures with a query generated in code.  We make use of temporary tables to run the query and also stash the matching Observation ids so we don't have to keep transferring them between the database and the server.
* Reuses the query to find the available Filters for the Choose Filters step (this will then be compatible when we start the multi-step Filter work) and so strips down the ObservationPredicateBuilder to only the Location parts - this can be completely decommissioned when the switch from Location codes to Location ids has been done - it's remaining to support the available Location search for the "Choose Locations" step.
* 
image

Update 02/02/22:
image

### CODE REVIEW

File: src/GovUk.Education.ExploreEducationStatistics.Data.Services/AbstractSubjectMetaService.cs [hunk 1]
```diff
@@ -12,12 +11,12 @@ namespace GovUk.Education.ExploreEducationStatistics.Data.Services
...
     public abstract class AbstractSubjectMetaService
     {
-        protected readonly IFilterItemRepository _filterItemRepository;
+        protected readonly IFilterItemRepository FilterItemRepository;
         protected static IComparer<string> LabelComparer { get; } = new LabelRelationalComparer();
 
         protected AbstractSubjectMetaService(IFilterItemRepository filterItemRepository)
         {
-            _filterItemRepository = filterItemRepository;
+            FilterItemRepository = filterItemRepository;
```
Comment (markhiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Data.Services/AbstractSubjectMetaService.cs:19]: Why the name change here?

**Review by markhiveit — CHANGES_REQUESTED** [review]

File: src/GovUk.Education.ExploreEducationStatistics.Data.Model/Database/StatisticsDbContext.cs [hunk 2]
```diff
@@ -1,16 +1,57 @@
...
+            if (ReferenceEquals(null, obj)) return false;
+            if (ReferenceEquals(this, obj)) return true;
+            if (obj.GetType() != this.GetType()) return false;
+            return Equals((IdTempTable)obj);
+        }
+
+        public override int GetHashCode()
+        {
+            return Id.GetHashCode();
+        }
```
[the remark this answers is not in the document]
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Model/Database/StatisticsDbContext.cs:49]: This is for convenience within tests.  When the code creates a temporary table and bulk inserts ids into it, we create a Moq expectation of the ids going into that table, and the expectations need these equality methods in order to provide a match.  It's possible to write a matcher without these methods in place too, but much easier to work with with the equality methods in place.  The MatchedObservations never get used when setting expectations, so they've never needed the equality methods.

File: src/GovUk.Education.ExploreEducationStatistics.Data.Services/ObservationService.cs [hunk 3]
```diff
@@ -33,227 +36,292 @@ public class ObservationService : AbstractRepository<Observation, long>, IObserv
...
+                    ? await GetLocationsClause(context, locationIds!, cancellationToken)
+                    : default;
+
+                var (filterItemIdsClause, filterItemIdTempTables) = !filterItemIds.IsNullOrEmpty()
+                    ? await GetSelectedFilterItemIdsClause(context, subjectId, filterItemIds!, cancellationToken)
+                    : default;
+
+                var sql =
+                    $"INSERT INTO #{nameof(MatchedObservation)} " +
+                    $"SELECT o.id FROM Observation o " +
```
[the remark this answers is not in the document]
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Services/ObservationService.cs:137]: Amended to use multiline string instead, as a little neater!

File: src/GovUk.Education.ExploreEducationStatistics.Data.Services/Interfaces/ITimePeriodService.cs [hunk 4]
```diff
@@ -3,17 +3,18 @@
 using System.Linq;
```
[the remark this answers is not in the document]
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Services/Interfaces/ITimePeriodService.cs:3]: Done

File: src/GovUk.Education.ExploreEducationStatistics.Data.Model/Repository/FilterItemRepository.cs [hunk 5]
```diff
@@ -37,17 +37,42 @@ public FilterItemRepository(StatisticsDbContext context) : base(context)
...
+                .SelectMany(filter => filter.FilterGroups)
+                .Select(filterGroup => filterGroup.Id);
+
+            var filterItems = await _context
+                .FilterItem
+                .AsNoTracking()
+                .Where(filterItem =>
+                    filterGroupIds.Contains(filterItem.FilterGroupId) &&
+                    _context.ObservationFilterItem.Any(ofi =>
+                        ofi.FilterItemId == filterItem.Id && matchedObservationIds.Contains(ofi.ObservationId)))
```
[the remark this answers is not in the document]
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Model/Repository/FilterItemRepository.cs:63]: Yeah it's pretty clever - I was kind've dumbstruck :)  Can't take credit for being the first to spot this though!

File: src/GovUk.Education.ExploreEducationStatistics.Data.Model/Repository/LocationRepository.cs [hunk 6]
```diff
@@ -39,57 +39,34 @@ public LocationRepository(StatisticsDbContext context)
...
+                .Where(o => o.SubjectId == subjectId);
 
-        public async Task<Dictionary<GeographicLevel, List<LocationAttributeNode>>> GetLocationAttributesHierarchicalByObservationsQuery(
-            IQueryable<Observation> observations,
-            Dictionary<GeographicLevel, List<string>>? hierarchies)
-        {
-            var locations = await observations
-                .AsNoTracking()
+            var locations = await
+                observations
```
[the remark this answers is not in the document]
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Model/Repository/LocationRepository.cs:52]: Done ta

File: src/GovUk.Education.ExploreEducationStatistics.Data.Model/Repository/LocationRepository.cs [hunk 7]
```diff
@@ -39,57 +39,34 @@ public LocationRepository(StatisticsDbContext context)
...
+            IList<Location> distinctLocations,
+            Dictionary<GeographicLevel, List<string>>? hierarchies = null)
         {
             var hierarchyWithLocationSelectors = hierarchies == null
                 ? new Dictionary<GeographicLevel, List<Func<Location, ILocationAttribute>>>()
                 : MapLocationAttributeSelectors(hierarchies);
 
-            return locations
+            return distinctLocations
+                .Distinct()
```
[the remark this answers is not in the document]
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Model/Repository/LocationRepository.cs:69]: I thought I'd put it there defensively.  I know the parameter name suggests to the calling code that it should be passing in a *distinct* list, but this'll ensure it!  Alternatively I could change this to be a check and to throw an exception if the list passed in isn't distinct, but this is I guess a bit riskier.

File: src/GovUk.Education.ExploreEducationStatistics.Data.Services/AbstractSubjectMetaService.cs (hunk 1 above)
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Services/AbstractSubjectMetaService.cs:19]: Rider's naming convention suggestion, as `_filterItemRepository` would make it a private variable, whereas in fact it's a protected variable.

**Review by markhiveit — APPROVED** [review]

File: src/GovUk.Education.ExploreEducationStatistics.Data.Services/LocationPredicateBuilder.cs [hunk 8]
```diff
@@ -8,62 +8,43 @@
...
             // TODO EES-3068 Migrate Location codes to ids in old Datablocks to remove this support for Location codes
-            if (query.Locations != null)
+            if (locationCodes != null)
             {
-                if (query.Locations.GeographicLevel != null)
+                if (locationCodes.GeographicLevel != null)
                 {
-                    predicate = predicate.AndAlso(observation =>
-                        observation.GeographicLevel == query.Locations.GeographicLevel);
+                    predicate = predicate.AndAlso(location => location.GeographicLevel == locationCodes.GeographicLevel);
```
Comment (benoutram) [inline src/GovUk.Education.ExploreEducationStatistics.Data.Services/LocationPredicateBuilder.cs:33]: Not really for this PR but I looked at all data blocks and permalinks and I don't think `GeographicLevel` has ever been allowed to be set as a query param so I think we could drop it in future.

File: src/GovUk.Education.ExploreEducationStatistics.Data.Model/Location.cs [hunk 9]
```diff
@@ -247,5 +247,95 @@ public PlanningArea PlanningArea
...
+            hashCode.Add(Region_Code);
+            hashCode.Add(Region_Name);
+            hashCode.Add(RscRegion_Code);
+            hashCode.Add(School_Code);
+            hashCode.Add(School_Name);
+            hashCode.Add(Sponsor_Code);
+            hashCode.Add(Sponsor_Name);
+            hashCode.Add(Ward_Code);
+            hashCode.Add(Ward_Name);
+            hashCode.Add(PlanningArea_Code);
```
Comment (benoutram) [inline src/GovUk.Education.ExploreEducationStatistics.Data.Model/Location.cs:336]: If we order these alphabetically both here and above it makes it a little easier to check the presence of them against a list of all the fields we expect.

File: src/GovUk.Education.ExploreEducationStatistics.Data.Services/SubjectMetaService.cs [hunk 10]
```diff
@@ -97,58 +107,74 @@ private async Task<SubjectMetaViewModel> GetSubjectMetaViewModel(Subject subject
...
+                    var locationQuery = LocationPredicateBuilder.Build(query.LocationIds, query.Locations);
+                    
+                    var locationIds = _context
+                        .Location
+                        .Where(locationQuery)
+                        .Select(l => l.Id);
+                    
+                    var observations = _context
+                        .Observation
+                        .AsNoTracking()
```
[the remark this answers is not in the document]
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Services/SubjectMetaService.cs:138]: Yup good idea - done

[the remark this answers is not in the document]
Comment (markhiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Services/SubjectMetaService.cs:138]: I saw this and assumed `locationIds` was still a queryable at this point, so adding `.AsNoTracking()` doesn't make a difference?

File: src/GovUk.Education.ExploreEducationStatistics.Data.Model/Location.cs (hunk 9 above)
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Model/Location.cs]: Moved Planning Area into the correct alphabetical place in Equals and GetHashCode methods, and in the order of fields themselves in Location.  Left GeographicLevel and Id where they were.

File: src/GovUk.Education.ExploreEducationStatistics.Data.Services/ResultSubjectMetaService.cs [hunk 11]
```diff
@@ -74,41 +74,44 @@ public class ResultSubjectMetaService : AbstractSubjectMetaService, IResultSubje
...
+                    var locations = observations
+                        .Select(o => o.Location)
+                        .Distinct()
+                        .ToList();
+                    
                     var locationAttributes =
-                        await _locationRepository.GetLocationAttributesHierarchicalByObservationsList(
-                            observations,
+                        _locationRepository.GetLocationAttributesHierarchical(
+                            locations,
```
[the remark this answers is not in the document]
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Services/ResultSubjectMetaService.cs]: Done.  Partially extracted out into extension methods / methods on Location class itself, and updated all tests to remove the use of now-removed Repo methods.

Also added tests around SubjectMetaService.GetSubjectMeta() method that didn't previously exist, for getting the list of available Time Periods and Filter Items for their respective steps in the Table Tool.

File: src/GovUk.Education.ExploreEducationStatistics.Data.Services/SubjectMetaService.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Services/SubjectMetaService.cs:142]: Could be Mark, I'm not sure actually!

**Review by benoutram — APPROVED** [review]

**Pull request merged**